Move full diagnostic rendering to `ruff_db`

ntBre opened the pull request:
## Summary

This PR switches the `full` output format in Ruff over to use the rendering code
in `ruff_db`. As proposed in the design doc, this involves a lot of
changes to the snapshot output.

I also had to comment out this assertion with a TODO to replace it after https://github.com/astral-sh/ruff/issues/19688 because many of Ruff's "file-level" annotations aren't actually file-level. They just happen to occur at the start of the file, especially in tests with very short snippets.

https://github.com/astral-sh/ruff/blob/529d81daca30e0af93a18d7406bd2c419008e8f4/crates/ruff_annotate_snippets/src/renderer/display_list.rs#L1204-L1208

I broke up the snapshot commits at the end into several blocks, but I don't think it's enough to help with review. The first few (notebooks, syntax errors, and test rules) are small enough to look at, but I couldn't really think of other categories beyond that. I'm happy to break those up or pick out specific examples beyond what I have below, if that would help.

The minimal code changes are in this [range](https://github.com/astral-sh/ruff/pull/19415/files/abd28f1e776e442379186e9f24c9aa6290f64163), with the snapshot commits following. Moving the `FullRenderer` and updating the `EmitterFlags` aren't strictly necessary either. I even dropped the renderer commit this morning but figured it made sense to keep it since we have the `full` module for tests. I don't feel strongly either way.

## Test Plan

I did actually click through all 1700 snapshots individually instead of
accepting them all at once, although I moved through them quickly. There are a
few main categories:

### Lint diagnostics

```diff
-unused.py:8:19: F401 [*] `pathlib` imported but unused
+F401 [*] `pathlib` imported but unused
+  --> unused.py:8:19
    |
  7 | # Unused, _not_ marked as required (due to the alias).
  8 | import pathlib as non_alias
-   |                   ^^^^^^^^^ F401
+   |                   ^^^^^^^^^
  9 |
 10 | # Unused, marked as required.
    |
-   = help: Remove unused import: `pathlib`
+help: Remove unused import: `pathlib`
```

- The filename and line numbers are moved to the second line
- The second noqa code next to the underline is removed

### Syntax errors

These are much like the above.

```diff
-    -:1:16: invalid-syntax: Expected one or more symbol names after import
+    invalid-syntax: Expected one or more symbol names after import
+     --> -:1:16
       |
     1 | from foo import
       |                ^
```

One thing I noticed while reviewing some of these, but I don't think is strictly syntax-error-related, is that some of the new diagnostics have a little less context after the error. I don't think this is a problem, but it's one small discrepancy I hadn't noticed before. Here's a minor example:

```diff
-syntax_errors.py:1:15: invalid-syntax: Expected one or more symbol names after import
+invalid-syntax: Expected one or more symbol names after import
+ --> syntax_errors.py:1:15
   |
 1 | from os import
   |               ^
 2 |
 3 | if call(foo
-4 |     def bar():
   |
```

And one of the biggest examples:

```diff
-E30_syntax_error.py:18:11: invalid-syntax: Expected ')', found newline
+invalid-syntax: Expected ')', found newline
+  --> E30_syntax_error.py:18:11
    |
 16 |         pass
 17 |
 18 | foo = Foo(
    |           ^
-19 |
-20 |
-21 | def top(
    |
```

Similarly, a few of the lint diagnostics showed that the cut indicator calculation for overly long lines is also slightly different, but I think that's okay too.

### Full-file diagnostics

```diff
-comment.py:1:1: I002 [*] Missing required import: `from __future__ import annotations`
+I002 [*] Missing required import: `from __future__ import annotations`
+--> comment.py:1:1
+help: Insert required import: `from __future__ import annotations`
+
```

As noted above, these will be much more rare after #19688 too. This case isn't a true full-file diagnostic and will render a snippet in the future, but you can see that we're now rendering the help message that would have been discarded before. In contrast, this is a true full-file diagnostic and should still look like this after #19688:

```diff
-__init__.py:1:1: A005 Module `logging` shadows a Python standard-library module
+A005 Module `logging` shadows a Python standard-library module
+--> __init__.py:1:1
```

### Jupyter notebooks

There's nothing particularly different about these, just showing off the cell index again.

```diff
-    Jupyter.ipynb:cell 3:1:7: F821 Undefined name `x`
+    F821 Undefined name `x`
+     --> Jupyter.ipynb:cell 3:1:7
       |
     1 | print(x)
-      |       ^ F821
+      |       ^
       |
```

MichaReiser commented:
Thank you for looking into this.

I'm not sure I understand what the diff shown for **Diagnostics with FixAvailability::Never** or **Sometimes** has to do with the fix availability. 

I think it would be more helpful to list the separate differences:

* The annotated message now shows the help text instead of the rule code (makes sense to me, but I haven't reviewed all snapshot changes)
* The line/column and filename moved: This makes sense to me and is a necessity to support multifile spans
* Full file diagnostics: @BurntSushi didn't you add some special casing for it to ruff's diagnostic rendering?

MichaReiser commented:
> These are the only ones I noticed that I wasn't very happy with. I tried to
address these by making the Range on Ruff's diagnostics None

We can't change the range or that would be breaking where suppression comments are allowed. We also can't remove the range or the diagnostic can no longer be suppressed. 


What I don't understand in your example. Is the rendered code frame just empty? Does it collapse lines or does it render the full file?

MichaReiser commented on crates/ruff_linter/src/rules/flake8_bugbear/snapshots/ruff_linter__rules__flake8_bugbear__tests__B911_B911.py.snap:
```diff
@@ -1,168 +1,168 @@
  8 | batched((foo for foo in cycle()))
  9 | batched(itertools.batched([1, 2, 3], strict=True))
-   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ B911
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Add an explicit `strict` parameter
```
I feel like the primary message looks very good in many cases but not all. 

For example here, I think the `help` text was better. I don't think this is something we need to change in rendering. Instead, we should change how the diagnostic is constructed in the rule. 

The airflow rules are similar. The inline texts are too long.

ntBre commented:
I think I'm getting pretty close to having the output we want, albeit with a potentially hacky implementation. For this input:

```python
import math

x: str = 1

match 42:
    case _: ...
```

and this shell command:

```shell
git diff <(ruff check --no-cache tr{i,y}.py --target-version py39) \
         <(myruff check --no-cache tr{i,y}.py --target-version py39)
```

I'm currently getting this diff:

```diff
-tri.py:1:1: E902 No such file or directory (os error 2)
-try.py:1:8: F401 [*] `math` imported but unused
+E902 No such file or directory (os error 2)
+
+F401 [*] `math` imported but unused
+ --> try.py:1:8
   |
 1 | import math
-  |        ^^^^ F401
+  |        ^^^^
 2 |
 3 | x: str = 1
   |
-  = help: Remove unused import: `math`
+help: Remove unused import: `math`
 
-try.py:5:1: SyntaxError: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)
+SyntaxError: Cannot use `match` statement on Python 3.9 (syntax was added in Python 3.10)
+ --> try.py:5:1
   |
 3 | x: str = 1
 4 |
```

This all looks good to me except the missing filename for the full-file diagnostic E902. I also tried a bit to get rid of the extra newline without much luck. I think the last two commits Andrew linked are in exactly the right place, but reverting to the version in [602a27c](https://github.com/astral-sh/ruff/commit/602a27c4e3861757e9d78257f49c58bb823d43b9) removes all the lines between diagnostics. I think I've convinced myself that it's actually nice to have the "extra" line for E902 since that's consistent with other diagnostics, but maybe I'm just giving up too easily.

Another issue with the check I added to remove the empty annotations is that some `TextRange::default` ranges are intentional, I think. For example, this snapshot changed:

https://github.com/astral-sh/ruff/blob/dc10ab81bd85b8af49ee2fe8ed4a6f767d37d6e0/crates/ty_ide/src/goto_declaration.rs#L289-L294

when I think it might want to point to the beginning of the file, with a real annotation. I002 is another example of a diagnostic that might actually want a default range and changed in my recent commits:

https://github.com/astral-sh/ruff/blob/dc10ab81bd85b8af49ee2fe8ed4a6f767d37d6e0/crates/ruff_linter/src/rules/isort/snapshots/ruff_linter__rules__isort__tests__required_import_with_alias_off.py.snap#L5-L8

I'll keep looking at these remaining issues tomorrow, but at least I made some progress on the brackets and blank lines (maybe) today.

BurntSushi commented:
I think adding a knob on to `Annotation` that lets you suppress code snippets. This "feels" better to me than trying to put a range on the diagnostic itself, which I think could be quite confusing. (Micha noted this.) While true we could enforce this with assertions, it feels like a more complicated change to the diagnostic model than just letting an annotation be "light" in the sense that there are no code snippets.

I do agree that this makes the name "annotation" somewhat of a misnomer in certain cases. Although, you could recast as, "it's still an annotation, but it applies to contents that are too large to usefully display in an error message."

MichaReiser commented:
Would it make sense to split this pr into 1) extending the rendering and b) migrating ruff?

Or can you organize the commits in a way that makes reviewing easier

ntBre commented:
This is up to date again, at least until we merge any other rule PRs, and I squashed the commits into a more linear history. I think I can still split more of them off into standalone PRs, though.

ntBre commented:
This is now based on #19653 and contains very few code changes, even fewer if I didn't split out the `FullRenderer`.

I just noticed that we aren't handling Jupyter notebook cell offsets like Ruff's old format or in the new `concise` format. So I guess I'll be splitting off another PR trying to get that information into `annotate-snippets` too. It seems like it could be a pretty invasive change without looking into it much yet.

MichaReiser commented:
> It seems we're no longer counting carriage returns as line breaks. This case is now reported as 1:40 instead of 2:1:

It's certainly unfortunate because it's a valid line terminator in Python. But I doubt that it's wildly used and it might be somewhat annoying to fix (unless we patch `\r` to `\n` in the source code normalization before passing it to annotated snippets)

> Claude called this an improvement, but I'm a bit skeptical. We now report 1:2 here instead of 1:1:

Zed does render the BOM character and also reports 1:2 for the `import`. So I guess it's not entirely unreasonable. I'm fine deferring this for now (or stripping the BOM as part of our source code normalization).


> It looks like annotate-snippets doesn't like our EOF ranges. These used to report a range of 30:1, the first character of the non-existent 30th line in this file), but now fall back to a default range of 1:1:

What do we use as range? Is it `TextRange::at(source_text.text_len(), 0)` or is it one beyond? If it's one beyond, thank I think the right thing is to fix the range. 

I think we need to fix this in some way that it doesn't point at the start of the file. It seems like this is only a bug in the code that determines the line number as the carret rendering seems correct.

ntBre commented:
(I'll split these off into a separate PR again too, this was just the easiest way to test them first)

ntBre commented:
This should be ready for review. I left the commits towards the end named as if they resolve the carriage return, BOM, and EOF issues, but after https://github.com/astral-sh/ruff/pull/19806 they just accept the snapshots from the actual code changes in the other PR.

I also wrote a much shorter AWK version of Claude's script for checking the offsets in the diff. It has three false positives, but I verified that they are just due to git displaying a larger diff in those areas, not real differences in `row:column` reporting.

<details><summary>Script and output</summary>
<p>

```awk
#!/usr/bin/awk -f

{
    # Old line like:
    # -    -:1:8: F401 [*] `os` imported but unused
    if (match($0, /-.*:([0-9]+):([0-9]+):/, m)) {
        row = m[1]
        col = m[2]
        line = $0
    }
    # New line like:
    # +    F401 [*] `os` imported but unused
    else if (match($0, /+.*-->.*:([0-9]+):([0-9]+)/, m)) {
        new_row = m[1]
        new_col = m[2]
        if (new_row != row || new_col != col) {
            print "MISMATCH:"
            printf "\tOld line: %s\n", line
            printf "\tNew line: %s\n", $0
        }
    }
}
```

These are false positives due to how git reports the diff for these files.

```shell
$ git diff main | ./check_offsets.awk
MISMATCH:
        Old line: -E501_1.py:5:89: E501 Line too long (150 > 88)
        New line: + --> E501_1.py:4:89
MISMATCH:
        Old line: -E501_1.py:6:89: E501 Line too long (149 > 88)
        New line: + --> E501_1.py:5:89
MISMATCH:
        Old line: -all.py:1:5: D103 Missing docstring in public function
        New line: +--> all.py:1:1
```

For example:

```diff
-all.py:1:1: D100 Missing docstring in public module
-all.py:1:5: D103 Missing docstring in public function
+D100 Missing docstring in public module
+--> all.py:1:1
+
+D103 Missing docstring in public function
+ --> all.py:1:5
```

The E501 case is similar. Basically all of the E501 diagnostics for the file are jumbled together in the diff because the long-line truncation changes too, but all of the individual diagnostics are still there and reported with the correct line and column numbers.

</p>
</details>

ntBre commented on crates/ruff_linter/src/message/text.rs:
```diff
@@ -94,23 +90,7 @@ impl Emitter for TextEmitter {
-            }
-
-            if self.flags.intersects(EmitterFlags::SHOW_FIX_DIFF) {
+            if self.show_fix_diff {
```
This is the diff rendering that only works in the snapshots (https://github.com/astral-sh/ruff/issues/7352). I didn't notice any changes in those sections of the snapshots, as expected, but is there something else you want me to check here?

MichaReiser replied:
It seems that `GroupEmitter::with_show_source` is only called in tests. So we could just delete it? But I also don't mind if you do this in a separate PR

MichaReiser replied:
Sounds good. I'm fine merging this PR and follow up in a separate PR.
